database.go

```diff
@@ -149,12 +163,104 @@ func (db *Database) AddBinary(binaryContent []byte) *Binary {
 	return db.Content.Meta.Binaries.Add(binaryContent, WithKDBXv31Binary)
 }
 
+type BinarySet map[*Binary]bool
+type BinaryDuplicates map[*Binary]*Binary
+type BinaryUsages map[int][]BinaryReference
+
+func (db *Database) CleanupBinaries() int {
+	binaries := Binaries{}
+	usages := db.getBinariesMap()
+	garbage := db.FindBinaryGarbage(usages)
+	duplicates := db.FindBinaryDuplicates()
+
+	for _, binary := range *db.GetBinaries() {
+		if _, ok := garbage[&binary]; ok {
+			continue
+		}
+		if original, ok := duplicates[&binary]; ok {
+			for _, ref := range usages[binary.ID] {
+				ref.Value.ID = original.ID
+			}
+			continue
+		}
+		binaries = append(binaries, binary)
+	}
+
+	counter := 0
+	for _, binary := range binaries {
+		for _, ref := range usages[binary.ID] {
+			binary.ID = counter
+			ref.Value.ID = binary.ID
+		}
+		counter++
+	}
+
+	*db.GetBinaries() = binaries
+	return len(garbage) + len(duplicates)
+}
+
+func (db *Database) FindBinaryDuplicates() BinaryDuplicates {
+	duplicates := BinaryDuplicates{}
+	original := BinarySet{}
+
+	for _, binary := range *db.GetBinaries() {
+		for originalBinary := range original {
+			if bytes.Equal(originalBinary.Content, binary.Content) {
+				duplicates[&binary] = originalBinary
+				break
+			}
+		}
+		original[&binary] = true
+	}
+
+	return duplicates
+}
+
+func (db *Database) FindBinaryGarbage(usages BinaryUsages) BinarySet {
+	garbage := BinarySet{}
+
+	for _, binary := range *db.GetBinaries() {
+		if 0 == len(usages[binary.ID]) {
+			garbage[&binary] = true
+		}
+	}
+
+	return garbage
+}
+
+func (db *Database) getBinariesMap() BinaryUsages {
```

Comment:
The name does not really reflect the intention. Especially if the result type is so clearly named.